### PR TITLE
Document "Custom Post" and "Custom Pre" Provisioning Snippets

### PR DESCRIPTION
**Pull request by maximiliankolb**
Documents https://github.com/theforeman/foreman/pull/9483

* [x] I am familiar with the [contributing](https://github.com/theforeman/foreman-documentation/blob/master/CONTRIBUTING.md) guidelines.

Please cherry-pick my commits into:

* [x] Foreman 3.5/Katello 4.7 (planned Satellite 6.13)
* [x] Foreman 3.4/Katello 4.6
* [x] Foreman 3.3/Katello 4.5 (Satellite 6.12)
* [x] Foreman 3.2/Katello 4.4
* [x] Foreman 3.1/Katello 4.3 (Satellite 6.11, orcharhino 6.1+)

### CODE REVIEW

File: guides/common/assembly_configuring-provisioning-resources.adoc [hunk 1]
```diff
@@ -50,6 +50,12 @@ include::modules/proc_creating-provisioning-templates.adoc[leveloffset=+1]
 
 include::modules/proc_cloning-provisioning-templates.adoc[leveloffset=+1]
 
+include::modules/proc_creating-custom-provisioning-hooks.adoc[leveloffset=+1]
+
+ifndef::satellite[]
```
Comment (sbernhard) [inline guides/common/assembly_configuring-provisioning-resources.adoc:55]: The custom snippets exists for satellite, too.

File: guides/common/modules/proc_creating-custom-provisioning-hooks.adoc [hunk 2]
```diff
@@ -0,0 +1,28 @@
+[id="Creating_Custom_Provisioning_Hooks_{context}"]
+= Creating Custom Provisioning Hooks
+
+A custom provisioning snippets allows you to execute custom code during host provisioning.
+You can run code before and/or after the provisioning process.
+
```
Comment (sbernhard) [inline guides/common/modules/proc_creating-custom-provisioning-hooks.adoc:6]: Please add a sentence like:
Depending on the original template, multiple custom snippet hooks exists which can be used to have custom snippets. Check the original template first to know, which custom snippet can be used.

File: guides/common/modules/proc_creating-custom-provisioning-hooks.adoc [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+.Procedure
+. In the {ProjectWebUI}, navigate to *Hosts* > *Provisioning Templates* and click *Create Template*.
+. In the *Name* field, enter a name for your custom provisioning snippet:
```
Comment (sbernhard) [inline guides/common/modules/proc_creating-custom-provisioning-hooks.adoc:9]: It should be clear, that the custom snippet hook need to exist. As said, there are many.

File: guides/common/modules/proc_creating-custom-provisioning-hooks.adoc [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+** Append ` custom pre` to the name of a provisioning template to run code before provisioning a host.
+** Append ` custom post` to the name of a provisioning template to run code after provisioning a host.
+. On the *Type* tab, select *Snippet*.
+. Click *Submit* to create your custom provisioning hook.
```
Comment (sbernhard) [inline guides/common/modules/proc_creating-custom-provisioning-hooks.adoc:13]: ... to create your custom provisining snippet.

File: guides/common/modules/proc_creating-custom-provisioning-hooks.adoc [hunk 5]
```diff
@@ -0,0 +1,28 @@
+[id="Creating_Custom_Provisioning_Hooks_{context}"]
+= Creating Custom Provisioning Hooks
```
Comment (sbernhard) [inline guides/common/modules/proc_creating-custom-provisioning-hooks.adoc:2]: Creating Custom Provisioning Snippets

File: guides/common/modules/ref_custom-provisioning-hooks-example.adoc [hunk 6]
```diff
@@ -0,0 +1,17 @@
+[id="Custom_Provisioning_Hooks_Example_{context}"]
+= Custom Provisioning Hooks Example
```
Comment (sbernhard) [inline guides/common/modules/ref_custom-provisioning-hooks-example.adoc:2]: Custom Provisioning Snippet Example

**Review by sbernhard — CHANGES_REQUESTED** [review]

File: guides/common/modules/ref_custom-provisioning-hooks-example.adoc (hunk 6 above)
Comment (maximiliankolb) [reply guides/common/modules/ref_custom-provisioning-hooks-example.adoc]: Renamed.

**Review by sbernhard — APPROVED** [review]

**Comment by maximiliankolb** [comment]
@adamlazik1 Would you mind taking a look at this? Thanks!

**Comment by maximiliankolb** [comment]
I have added an example for managed hosts running EL.

File: guides/common/modules/ref_custom-provisioning-snippet-example-for-enterprise-linux.adoc [hunk 7]
```diff
@@ -0,0 +1,17 @@
+[id="Custom_Provisioning_Snippet_Example_for_Enterprise_Linux_{context}"]
+= Custom Provisioning Snippet Example for {EL}
+
+You can use `Custom Post` snippets to call external APIs from within the provisioning template directly after provisioning a hosts.
+
+.Custom Post Hook Example for {EL}
```
Comment (sbernhard) [inline guides/common/modules/ref_custom-provisioning-snippet-example-for-enterprise-linux.adoc:6]: Should we put the name of the custom snippet to make it clear where the snippet is added?

File: guides/common/modules/ref_custom-provisioning-snippet-example-for-debian.adoc [hunk 8]
```diff
@@ -0,0 +1,17 @@
+[id="Custom_Provisioning_Snippet_Example_for_Debian_{context}"]
+= Custom Provisioning Snippet Example for Debian
+
+You can use `Custom Post` snippets to call external APIs from within the provisioning template directly after provisioning a hosts.
+
+.Custom Post Hook Example for Debian/Ubuntu
```
Comment (sbernhard) [inline guides/common/modules/ref_custom-provisioning-snippet-example-for-debian.adoc:6]: Should we put the name of the custom snippet to make it clear where the snippet is added?

File: guides/common/modules/ref_custom-provisioning-snippet-example-for-enterprise-linux.adoc (hunk 7 above)
Comment (maximiliankolb) [reply guides/common/modules/ref_custom-provisioning-snippet-example-for-enterprise-linux.adoc]: Added.

File: guides/common/modules/ref_custom-provisioning-snippet-example-for-debian.adoc (hunk 8 above)
Comment (maximiliankolb) [reply guides/common/modules/ref_custom-provisioning-snippet-example-for-debian.adoc]: Added.

File: guides/common/modules/proc_creating-custom-provisioning-snippets.adoc [hunk 9]
```diff
@@ -0,0 +1,33 @@
+[id="Creating_Custom_Provisioning_Snippets_{context}"]
+= Creating Custom Provisioning Snippets
+
+A custom provisioning snippets allows you to execute custom code during host provisioning.
```
Comment (adamlazik1) [inline guides/common/modules/proc_creating-custom-provisioning-snippets.adoc:4]:
```suggestion
Custom provisioning snippets allow you to execute custom code during host provisioning.
```

File: guides/common/modules/proc_creating-custom-provisioning-snippets.adoc [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+You can run code before and/or after the provisioning process.
+
+.Prerequisite
+* Depending on your provisioning template, multiple custom snippet hooks exist which you can use to include custom provisioning snippets.
+Ensure that you check your provisioning template first to verify which custom snippet you can use.
```
Comment (adamlazik1) [inline guides/common/modules/proc_creating-custom-provisioning-snippets.adoc:9]:
```suggestion
Ensure that you check your provisioning template first to verify which custom snippets you can use.
```

File: guides/common/modules/proc_creating-custom-provisioning-snippets.adoc [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+.Procedure
+. In the {ProjectWebUI}, navigate to *Hosts* > *Provisioning Templates* and click *Create Template*.
+. In the *Name* field, enter a name for your custom provisioning snippet.
+The name has to start with the name of a provisioning template that supports including custom provisioning snippets:
+** Append ` custom pre` to the name of a provisioning template to run code before provisioning a host.
+** Append ` custom post` to the name of a provisioning template to run code after provisioning a host.
```
Comment (adamlazik1) [inline guides/common/modules/proc_creating-custom-provisioning-snippets.adoc:16]: The introduction says you can run code before **and/or** after the provisioning, so is there an option for both before and after? Or do I just append both?

File: guides/common/modules/ref_custom-provisioning-snippet-example-for-debian.adoc [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+[options="nowrap" subs="+quotes"]
+----
+echo "Calling API to report successful host deployment"
+
+apt-get install -y curl ca-certificates
+
+curl -X POST \
+     -H "Content-Type: application/json" \
+     -d '{"name": "<%= @host.name %>", "operating_system": "<%= @host.operatingsystem.name %>", "status": "provisioned",}' \
+     "https://api.example.com/"
```
Comment (adamlazik1) [inline guides/common/modules/ref_custom-provisioning-snippet-example-for-debian.adoc:16]:
```suggestion
-H "Content-Type: application/json" \
-d '{"name": "<%= @host.name %>", "operating_system": "<%= @host.operatingsystem.name %>", "status": "provisioned",}' \
"https://api.example.com/"
```

File: guides/common/modules/ref_custom-provisioning-snippet-example-for-debian.adoc [hunk 13]
```diff
@@ -0,0 +1,17 @@
+[id="Custom_Provisioning_Snippet_Example_for_Debian_{context}"]
+= Custom Provisioning Snippet Example for Debian
+
+You can use `Custom Post` snippets to call external APIs from within the provisioning template directly after provisioning a hosts.
```
Comment (adamlazik1) [inline guides/common/modules/ref_custom-provisioning-snippet-example-for-debian.adoc:4]:
```suggestion
You can use `Custom Post` snippets to call external APIs from within the provisioning template directly after provisioning a host.
```

File: guides/common/modules/ref_custom-provisioning-snippet-example-for-enterprise-linux.adoc [hunk 14]
```diff
@@ -0,0 +1,17 @@
+[id="Custom_Provisioning_Snippet_Example_for_Enterprise_Linux_{context}"]
+= Custom Provisioning Snippet Example for {EL}
+
+You can use `Custom Post` snippets to call external APIs from within the provisioning template directly after provisioning a hosts.
```
Comment (adamlazik1) [inline guides/common/modules/ref_custom-provisioning-snippet-example-for-enterprise-linux.adoc:4]:
```suggestion
You can use `Custom Post` snippets to call external APIs from within the provisioning template directly after provisioning a host.
```

File: guides/common/modules/ref_custom-provisioning-snippet-example-for-enterprise-linux.adoc [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+[options="nowrap" subs="+quotes"]
+----
+echo "Calling API to report successful host deployment"
+
+yum install -y curl ca-certificates
+
+curl -X POST \
+     -H "Content-Type: application/json" \
+     -d '{"name": "<%= @host.name %>", "operating_system": "<%= @host.operatingsystem.name %>", "status": "provisioned",}' \
+     "https://api.example.com/"
```
Comment (adamlazik1) [inline guides/common/modules/ref_custom-provisioning-snippet-example-for-enterprise-linux.adoc:16]:
```suggestion
-H "Content-Type: application/json" \
-d '{"name": "<%= @host.name %>", "operating_system": "<%= @host.operatingsystem.name %>", "status": "provisioned",}' \
"https://api.example.com/"
```

**Review by adamlazik1 — COMMENTED** [review]
LGTM, only some small notes.

File: guides/common/modules/proc_creating-custom-provisioning-snippets.adoc (hunk 9 above)
Comment (maximiliankolb) [reply guides/common/modules/proc_creating-custom-provisioning-snippets.adoc:4]: Applied.

File: guides/common/modules/proc_creating-custom-provisioning-snippets.adoc (hunk 10 above)
Comment (maximiliankolb) [reply guides/common/modules/proc_creating-custom-provisioning-snippets.adoc:9]: applied.

File: guides/common/modules/proc_creating-custom-provisioning-snippets.adoc (hunk 11 above)
Comment (maximiliankolb) [reply guides/common/modules/proc_creating-custom-provisioning-snippets.adoc:16]: Yes, you can use both if you want to, but with separate snippets.

File: guides/common/modules/ref_custom-provisioning-snippet-example-for-debian.adoc (hunk 13 above)
Comment (maximiliankolb) [reply guides/common/modules/ref_custom-provisioning-snippet-example-for-debian.adoc:4]: Applied.

File: guides/common/modules/ref_custom-provisioning-snippet-example-for-debian.adoc (hunk 12 above)
Comment (maximiliankolb) [reply guides/common/modules/ref_custom-provisioning-snippet-example-for-debian.adoc:16]: Applied.

File: guides/common/modules/ref_custom-provisioning-snippet-example-for-enterprise-linux.adoc (hunk 14 above)
Comment (maximiliankolb) [reply guides/common/modules/ref_custom-provisioning-snippet-example-for-enterprise-linux.adoc:4]: Applied.

**Review by maximiliankolb — COMMENTED** [review]
Rebased to master and applied all feedback. Merging after GHA is green.

**Comment by maximiliankolb** [comment]
Thanks Adam for reviewing my PR!

Merged to master and cherry-picked: :heavy_check_mark: 
   f8585eac2..1015c4e14  3.5 -> 3.5
   dbe9896cd..7e171ee18  3.4 -> 3.4
   41ed95c66..c078a3f3b  3.3 -> 3.3

**Pull request merged**